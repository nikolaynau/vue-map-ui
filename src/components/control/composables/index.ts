export * from './injectionSymbols';
export * from './useLayersControl';
export * from './useZoomControl';
export * from './useScaleControl';
export * from './useAttributionControl';
