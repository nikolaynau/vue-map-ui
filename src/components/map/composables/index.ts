export * from './injectionSymbols';
export * from './useMap';
export * from './usePane';
