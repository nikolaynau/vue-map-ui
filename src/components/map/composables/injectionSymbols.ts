import type { InjectionKey, Ref } from 'vue';
import type { Map } from 'leaflet';

export const mapKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'Map'
    : ''
) as InjectionKey<Ref<Map | null>>;

export const paneKey = Symbol(
  import.meta.env.MODE === 'development' || import.meta.env.MODE === 'test'
    ? 'Pane'
    : ''
) as InjectionKey<Ref<HTMLElement | null>>;
