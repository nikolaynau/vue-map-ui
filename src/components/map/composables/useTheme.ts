import { computed, type ComputedRef, type MaybeRefOrGetter } from 'vue';
import { toRef, usePreferredDark } from '@vueuse/core';

export interface UseThemeOptions {
  cssPrefix?: MaybeRefOrGetter<string>;
}

export type UseThemeReturn = ComputedRef<string>;

export function useTheme(
  name: MaybeRefOrGetter<string | 'auto' | null | undefined>,
  options: UseThemeOptions = {}
): UseThemeReturn {
  const { cssPrefix = 'v-map-theme--' } = options;

  const _name = toRef(name);
  const _cssPrefix = toRef(cssPrefix);
  const _isDark = usePreferredDark();

  return computed(() => {
    return `${_cssPrefix.value}${
      !_name.value || _name.value === 'auto'
        ? _isDark.value
          ? 'dark'
          : 'light'
        : _name.value
    }`;
  });
}
