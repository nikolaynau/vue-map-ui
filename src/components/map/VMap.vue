<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
  inheritAttrs: false
});
</script>

<script setup lang="ts">
import { ref, toRefs } from 'vue';
import type {
  MapOptions,
  LatLngBoundsExpression,
  LatLngExpression,
  LeafletEventHandlerFn
} from 'leaflet';
import {
  useLeafletMap,
  useLeafletReady,
  type ViewChangedEvent
} from 'vue-use-leaflet';
import { provideMap } from './composables';
import { useAttrs, useEvents } from '../../composables';
import { camelizeKeys, omit, pick } from '../../utils/objects';

export interface Props {
  center?: LatLngExpression;
  zoom?: number;
  bounds?: LatLngBoundsExpression;
  useFly?: boolean;
  elementAttrs?: string[];
}

export type Attrs = MapOptions;

const props = withDefaults(defineProps<Props>(), {
  center: () => [0, 0],
  zoom: 0,
  bounds: undefined,
  useFly: false,
  elementAttrs: undefined
});

const { center, zoom, bounds, useFly } = toRefs(props);
const container = ref<HTMLElement | null>(null);
const { events, attrs } = useAttrs<LeafletEventHandlerFn>(false);
const _elementAttrs = [
  ...(props.elementAttrs ?? []),
  ...['id', 'class', 'style']
];
const leafletEvents = omit(events, ['viewChanged']);
const leafletOptions = camelizeKeys(omit(attrs, _elementAttrs));
const elementAttrs = pick(attrs, _elementAttrs);
const onViewChanged = events['viewChanged'] as (e: ViewChangedEvent) => void;

const map = useLeafletMap(container, {
  center,
  zoom,
  bounds,
  useFly,
  ...leafletOptions,
  onViewChanged
});
const ready = useLeafletReady(map);
useEvents(map, leafletEvents);
provideMap(map);

defineExpose({
  container,
  map
});
</script>

<template>
  <div v-bind="elementAttrs" class="v-map" ref="container">
    <slot v-if="ready"></slot>
  </div>
</template>

<style>
@import 'map';
</style>
