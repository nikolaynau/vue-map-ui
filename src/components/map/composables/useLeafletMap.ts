import { computed, markRaw, ref, unref, watch, type Ref } from 'vue';
import { isDef, tryOnUnmounted, type MaybeRef } from '@vueuse/shared';
import {
  latLngBounds,
  latLng,
  Map,
  type LatLngBoundsExpression,
  type LatLngExpression,
  type MapOptions,
  type LeafletEvent,
  type LatLng,
  type LatLngBounds
} from 'leaflet';
import { useLeafletEvent } from '@/composables/useLeafletEvent';

export interface UseLeafletMapOptions extends UseLeafletMapCallbacks {
  center?: MaybeRef<LatLngExpression | undefined>;
  zoom?: MaybeRef<number | undefined>;
  bounds?: MaybeRef<LatLngBoundsExpression | undefined>;
  useFly?: MaybeRef<boolean | undefined>;
  leafletOptions?: MapOptions;
  events?: Array<string>;
}

export interface UseLeafletMapCallbacks {
  onEvent?: (ev: LeafletEvent) => void;
  onViewChanged?: (ev: ViewChangedEvent) => void;
}

export interface ViewChangedEvent extends LeafletEvent {
  center: LatLng;
  zoom: number;
  bounds: LatLngBounds;
}

const DEFAULT_OPTIONS: Readonly<UseLeafletMapOptions> = {
  center: [0, 0],
  zoom: 0,
  useFly: false
};

export function useLeafletMap(
  element: MaybeRef<HTMLElement | null | undefined>,
  options: UseLeafletMapOptions = {}
): Ref<Map | null> {
  const {
    center = DEFAULT_OPTIONS.center,
    zoom = DEFAULT_OPTIONS.zoom,
    useFly = DEFAULT_OPTIONS.useFly,
    leafletOptions = {},
    bounds,
    events,
    onEvent,
    onViewChanged
  } = options;

  const map = ref<Map | null>(null) as Ref<Map | null>;
  const view = computed(() => ({ center: unref(center), zoom: unref(zoom) }));

  function initializeMap(element: HTMLElement) {
    const needInitView = unref(bounds) || unref(useFly);

    if (needInitView) {
      leafletOptions.center = undefined;
      leafletOptions.zoom = undefined;
    } else {
      if (!isDef(leafletOptions.center)) {
        leafletOptions.center = unref(center);
      }
      if (!isDef(leafletOptions.zoom)) {
        leafletOptions.zoom = unref(zoom);
      }
    }

    const mapInstance = new Map(element, leafletOptions);

    if (needInitView) {
      setInitialView(mapInstance);
    }

    map.value = markRaw(mapInstance);
  }

  function setInitialView(map: Map) {
    if (unref(bounds)) {
      if (unref(useFly)) {
        map
          .setView(
            DEFAULT_OPTIONS.center as LatLngExpression,
            DEFAULT_OPTIONS.zoom as number
          )
          .flyToBounds(unref(bounds)!);
      } else {
        map.fitBounds(unref(bounds)!);
      }
    } else {
      if (unref(useFly)) {
        map
          .setView(
            DEFAULT_OPTIONS.center as LatLngExpression,
            DEFAULT_OPTIONS.zoom as number
          )
          .flyTo(unref(center)!, unref(zoom));
      }
    }
  }

  function setBounds(bounds: LatLngBoundsExpression) {
    bounds = latLngBounds(bounds as LatLngExpression[]);
    if (!bounds.isValid()) {
      return;
    }

    if (unref(useFly)) {
      map.value?.flyToBounds(bounds);
    } else {
      map.value?.fitBounds(bounds);
    }
  }

  function setView(center: LatLngExpression, zoom?: number) {
    if (unref(useFly)) {
      map.value?.flyTo(center, zoom);
    } else {
      map.value?.setView(center, zoom);
    }
  }

  function setCenter(center: LatLngExpression) {
    if (unref(useFly)) {
      map.value?.flyTo(center);
    } else {
      map.value?.panTo(center);
    }
  }

  function setZoom(zoom: number) {
    if (unref(useFly)) {
      map.value?.flyTo(map.value?.getCenter(), zoom);
    } else {
      map.value?.setZoom(zoom);
    }
  }

  function latLngEquals(
    latLngA: LatLngExpression | null | undefined,
    latLngB: LatLngExpression | null | undefined
  ): boolean {
    if (latLngA === latLngB) {
      return true;
    }
    if (!isDef(latLngA) || !isDef(latLngB)) {
      return false;
    }
    return latLng(latLngA!).equals(latLng(latLngB!));
  }

  function destroyMap() {
    map.value?.off().remove();
    map.value = null;
  }

  watch(
    () => unref(element),
    value => {
      destroyMap();
      if (value) {
        initializeMap(value);
      }
    },
    {
      immediate: true,
      flush: 'sync'
    }
  );

  watch(
    () => unref(bounds),
    value => {
      if (value) {
        setBounds(value);
      }
    }
  );

  watch(view, (newValue, oldValue) => {
    if (
      !latLngEquals(newValue.center, oldValue.center) &&
      newValue.zoom !== oldValue.zoom
    ) {
      isDef(newValue.center) && setView(newValue.center, newValue.zoom);
    } else if (!latLngEquals(newValue.center, oldValue.center)) {
      isDef(newValue.center) && setCenter(newValue.center);
    } else if (newValue.zoom !== oldValue.zoom) {
      isDef(newValue.zoom) && setZoom(newValue.zoom);
    }
  });

  if (events && onEvent && events.length > 0) {
    useLeafletEvent(map, events, onEvent);
  }

  if (onViewChanged) {
    useLeafletEvent(map, 'moveend', (ev: LeafletEvent) => {
      const map = ev.sourceTarget as Map;
      onViewChanged({
        center: map.getCenter(),
        zoom: map.getZoom(),
        bounds: map.getBounds(),
        ...ev
      });
    });
  }

  tryOnUnmounted(() => {
    destroyMap();
  });

  return map;
}

export type UseLeafletMapReturn = ReturnType<typeof useLeafletMap>;
