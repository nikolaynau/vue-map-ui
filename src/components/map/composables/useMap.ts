import type { Map } from 'leaflet';
import { inject, provide, type Ref } from 'vue';
import { mapKey } from '../utils/injectionSymbols';

export function provideMap(map: Ref<Map | null>) {
  provide(mapKey, map);
}

export function useMap(): Ref<Map | null> {
  return inject(mapKey, undefined) as Ref<Map | null>;
}
