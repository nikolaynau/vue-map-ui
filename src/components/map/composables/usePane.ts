import { inject, provide, readonly, type Ref } from 'vue';
import { paneKey } from './injectionSymbols';

export function providePane(pane: Ref<HTMLElement | null>) {
  provide(paneKey, readonly(pane) as Ref<HTMLElement | null>);
}

export function usePane(): Ref<HTMLElement | null> | undefined {
  return inject(paneKey, undefined);
}
