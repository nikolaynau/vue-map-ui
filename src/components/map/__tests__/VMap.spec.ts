import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import VMap from '../VMap.vue';
import { h, type Ref, nextTick } from 'vue';
import { useMap } from '../composables/useMap';
import { LatLngBounds, type Map, type LatLngBoundsExpression } from 'leaflet';
import type { ViewChangedEvent } from '../composables/useLeafletMap';

describe('VMap', () => {
  it('render default props', () => {
    const wrapper = mount(VMap);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('map and container exist after component is mounted', () => {
    const wrapper = mount(VMap);
    expect(wrapper.vm.container).toBeDefined();
    expect(wrapper.vm.map).toBeDefined();
  });

  it('center and zoom props', () => {
    const wrapper = mount(VMap, { props: { center: [1, 2], zoom: 3 } });
    expect(wrapper.vm.map?.getZoom()).toBe(3);
    expect(wrapper.vm.map?.getCenter()).toEqual({ lat: 1, lng: 2 });
  });

  it('bounds prop', () => {
    const expected: LatLngBoundsExpression = [
      [1, 2],
      [3, 4]
    ];
    const wrapper = mount(VMap, { props: { bounds: expected } });
    expect(wrapper.vm.map?.getBounds()).toBeInstanceOf(LatLngBounds);
  });

  it('inherit css class', () => {
    const wrapper = mount(VMap, { attrs: { class: 'some-class' } });
    expect(wrapper.find('.v-map').classes('some-class')).toBe(true);
  });

  it('inherit css style', () => {
    const expected = 'height: 100px';
    const wrapper = mount(VMap, { attrs: { style: expected } });
    expect(wrapper.find('.v-map').attributes('style')).contains(expected);
  });

  it('inherit id attrs', () => {
    const expected = 'map-1';
    const wrapper = mount(VMap, { attrs: { id: expected } });
    expect(wrapper.find('.v-map').attributes('id')).contains(expected);
  });

  it('inherit any attrs', () => {
    const expected = 'map-1';
    const wrapper = mount(VMap, {
      props: { excludeAttrs: ['data-id'] },
      attrs: { 'data-id': expected }
    });
    expect(wrapper.find('.v-map').attributes('data-id')).contains(expected);
  });

  it('set leaflet property', () => {
    const wrapper = mount(VMap, {
      attrs: { maxZoom: 5 }
    });
    expect(wrapper.vm.map?.getMaxZoom()).toBe(5);
  });

  it('default slot', () => {
    const wrapper = mount(VMap, {
      slots: {
        default: {
          setup() {
            return () => h('div', { class: 'child-component' });
          }
        }
      }
    });

    expect(wrapper.find('.child-component').exists()).toBe(true);
  });

  it('provide map', () => {
    let map: Readonly<Ref<Map | null>> | undefined = undefined;
    const wrapper = mount(VMap, {
      slots: {
        default: {
          setup() {
            map = useMap();
            return () => h('div', { class: 'child' });
          }
        }
      }
    });

    expect(map).toBeDefined();
    expect(wrapper.vm.map).toBeDefined();
    expect(wrapper.vm.map).toBe(map!.value);
  });

  it('leaflet event', async () => {
    const moveendListener = vi.fn();
    const zoomendListener = vi.fn();

    const wrapper = mount(VMap, {
      attrs: {
        onMoveend: moveendListener,
        onZoomend: zoomendListener
      }
    });

    await nextTick();

    wrapper.vm.map?.setView([1, 2], 3, { animate: false });

    expect(moveendListener).toBeCalledTimes(1);
    expect(zoomendListener).toBeCalledTimes(1);
  });

  it('view changed event', async () => {
    const listener = vi.fn();
    const wrapper = mount(VMap, {
      attrs: {
        onViewChanged: listener
      }
    });

    await nextTick();

    wrapper.vm.map?.setView([1, 2], 3, { animate: false });

    expect(listener).toBeCalledTimes(1);
    const ev = listener.mock.calls[0][0] as ViewChangedEvent;
    expect(ev.center).toEqual({ lat: 1, lng: 2 });
    expect(ev.zoom).toBe(3);
    expect(ev.bounds).toBeInstanceOf(LatLngBounds);
  });
});
