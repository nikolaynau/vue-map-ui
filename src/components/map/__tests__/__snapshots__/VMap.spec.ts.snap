// Vitest Snapshot v1

exports[`VMap > render default props 1`] = `"<div class=\\"v-map leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom\\" style=\\"position: relative;\\" tabindex=\\"0\\"><!--v-if--><div class=\\"leaflet-pane leaflet-map-pane\\" style=\\"left: 0px; top: 0px;\\"><div class=\\"leaflet-pane leaflet-tile-pane\\"></div><div class=\\"leaflet-pane leaflet-overlay-pane\\"></div><div class=\\"leaflet-pane leaflet-shadow-pane\\"></div><div class=\\"leaflet-pane leaflet-marker-pane\\"></div><div class=\\"leaflet-pane leaflet-tooltip-pane\\"></div><div class=\\"leaflet-pane leaflet-popup-pane\\"></div></div><div class=\\"leaflet-control-container\\"><div class=\\"leaflet-top leaflet-left\\"><div class=\\"leaflet-control-zoom leaflet-bar leaflet-control\\"><a class=\\"leaflet-control-zoom-in\\" href=\\"#\\" title=\\"Zoom in\\" role=\\"button\\" aria-label=\\"Zoom in\\" aria-disabled=\\"false\\"><span aria-hidden=\\"true\\">+</span></a><a class=\\"leaflet-control-zoom-out leaflet-disabled\\" href=\\"#\\" title=\\"Zoom out\\" role=\\"button\\" aria-label=\\"Zoom out\\" aria-disabled=\\"true\\"><span aria-hidden=\\"true\\">âˆ’</span></a></div></div><div class=\\"leaflet-top leaflet-right\\"></div><div class=\\"leaflet-bottom leaflet-left\\"></div><div class=\\"leaflet-bottom leaflet-right\\"><div class=\\"leaflet-control-attribution leaflet-control\\"><a href=\\"https://leafletjs.com\\" title=\\"A JavaScript library for interactive maps\\">Leaflet</a></div></div></div></div>"`;
